### PR TITLE
Feature/chat backend

### PR DESCRIPTION
**Pull request by florianlenz**
- added chat backend based on our specs

### CODE REVIEW

File: backend/auth.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+			return nil, errors.New("got invalid amount of toSign bytes")
+		}
+
+		myBytes := make([]byte, 4)
+		_, err := rand.Read(myBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		toSign := append(myBytes, auth.ToSign...)
```
Comment (borjantrajanoski) [inline backend/auth.go]: Instinctively, I would switch the order here, as usually then we append something we add something to the right side of the original data. In this way the backend can use bytes.HasPrefix()

`toSign := append(auth.ToSign, myBytes...)`

File: backend/backend.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+				})
+				if err != nil {
+					return err
+				}
+
+			}
+		}
+
+		// handle responses
+		if msg.Response != nil {
```
Comment (borjantrajanoski) [inline backend/backend.go:111]: By just handling `if msg.Response != nil`, an invalid message containing both a Request and a Response would still be handled here. 

Should add a check similar to 
```
if msg.Request != nil && if msg.Response != nil {
    return errors.New("A message can’t have a response and a request at the same time.")
}
```

File: backend/auth.go (hunk 1 above)
Comment (florianlenz) [reply backend/auth.go]: Make sense. I switched it

**Review by borjantrajanoski — CHANGES_REQUESTED** [review]

File: backend/backend.go (hunk 2 above)
Comment (florianlenz) [reply backend/backend.go:111]: thanks. Added the check

**Review by borjantrajanoski — APPROVED** [review]

**Pull request merged**